File: jwt_decoder.py
# jwt_decoder.py
import sys
import base64
import json
import argparse
import re

# ANSI-цвета
COLOR_RESET = "\033[0m"
COLOR_KEY = "\033[94m"      # синий
COLOR_STRING = "\033[92m"   # зелёный
COLOR_NUMBER = "\033[93m"   # жёлтый
COLOR_BOOL = "\033[95m"     # пурпурный
COLOR_NULL = "\033[90m"     # серый
COLOR_HEADER = "\033[96m"   # голубой
COLOR_SIGNATURE = "\033[90m" # тёмно-серый

def base64url_decode(data):
    """Декодирует Base64Url в строку."""
    # Добавляем padding, если необходимо
    padding = 4 - (len(data) % 4)
    if padding != 4:
        data += '=' * padding
    return base64.urlsafe_b64decode(data).decode('utf-8')

def colorize_json(obj, indent=2, level=0):
    """Рекурсивно форматирует JSON с подсветкой."""
    if isinstance(obj, dict):
        lines = []
        if level == 0:
            lines.append("{")
        else:
            lines.append("{")
        for i, (k, v) in enumerate(obj.items()):
            key_str = f"{COLOR_KEY}\"{k}\"{COLOR_RESET}: "
            if isinstance(v, dict):
                val_str = colorize_json(v, indent, level+1)
            elif isinstance(v, list):
                val_str = colorize_json(v, indent, level+1)
            elif isinstance(v, str):
                val_str = f"{COLOR_STRING}\"{v}\"{COLOR_RESET}"
            elif isinstance(v, bool):
                val_str = f"{COLOR_BOOL}{str(v).lower()}{COLOR_RESET}"
            elif v is None:
                val_str = f"{COLOR_NULL}null{COLOR_RESET}"
            elif isinstance(v, (int, float)):
                val_str = f"{COLOR_NUMBER}{v}{COLOR_RESET}"
            else:
                val_str = str(v)
            lines.append(" " * (level+1) * indent + key_str + val_str)
            if i < len(obj)-1:
                lines[-1] += ","
        if level == 0:
            lines.append("}")
        else:
            lines.append(" " * level * indent + "}")
        return "\n".join(lines)
    elif isinstance(obj, list):
        lines = ["["]
        for item in obj:
            if isinstance(item, dict):
                lines.append(colorize_json(item, indent, level+1))
            else:
                lines.append(" " * (level+1) * indent + colorize_json(item, indent, level+1))
        lines.append(" " * level * indent + "]")
        return "\n".join(lines)
    else:
        if isinstance(obj, str):
            return f"{COLOR_STRING}\"{obj}\"{COLOR_RESET}"
        elif isinstance(obj, bool):
            return f"{COLOR_BOOL}{str(obj).lower()}{COLOR_RESET}"
        elif obj is None:
            return f"{COLOR_NULL}null{COLOR_RESET}"
        elif isinstance(obj, (int, float)):
            return f"{COLOR_NUMBER}{obj}{COLOR_RESET}"
        return str(obj)

def decode_jwt(token, pretty=True):
    """Декодирует JWT и возвращает header, payload, signature."""
    parts = token.split('.')
    if len(parts) != 3:
        raise ValueError("Неверный JWT: должно быть 3 сегмента")
    header_b64, payload_b64, signature_b64 = parts
    try:
        header_json = base64url_decode(header_b64)
        payload_json = base64url_decode(payload_b64)
        header = json.loads(header_json)
        payload = json.loads(payload_json)
    except Exception as e:
        raise ValueError(f"Ошибка декодирования Base64: {e}")
    return header, payload, signature_b64

def main():
    parser = argparse.ArgumentParser(description="JWT Decoder с подсветкой JSON")
    parser.add_argument("token", nargs="?", help="JWT-токен")
    parser.add_argument("--file", "-f", help="Читать токен из файла")
    parser.add_argument("--compact", action="store_true", help="Вывод без отступов")
    args = parser.parse_args()

    token = args.token
    if args.file:
        with open(args.file, 'r') as f:
            token = f.read().strip()
    if not token:
        print("Не указан JWT-токен.", file=sys.stderr)
        sys.exit(1)

    try:
        header, payload, signature = decode_jwt(token)
    except Exception as e:
        print(f"Ошибка: {e}", file=sys.stderr)
        sys.exit(1)

    indent = None if args.compact else 2
    print(f"{COLOR_HEADER}--- Header ---{COLOR_RESET}")
    print(colorize_json(header, indent=indent if indent else 0))
    print(f"{COLOR_HEADER}--- Payload ---{COLOR_RESET}")
    print(colorize_json(payload, indent=indent if indent else 0))
    print(f"{COLOR_SIGNATURE}--- Signature ---{COLOR_RESET}")
    print(f"{COLOR_SIGNATURE}{signature}{COLOR_RESET}")

if __name__ == "__main__":
    main()

File: jwt_decoder.rs
// jwt_decoder.rs
use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use base64::Engine;
use serde_json::Value;
use std::fs;
use std::env;
use termcolor::{Color, ColorChoice, ColorSpec, StandardStream, WriteColor};
use std::io::Write;

const RESET: &str = "\x1b[0m";
const KEY: &str = "\x1b[94m";
const STR: &str = "\x1b[92m";
const NUM: &str = "\x1b[93m";
const BOOL: &str = "\x1b[95m";
const NULL: &str = "\x1b[90m";
const HEADER: &str = "\x1b[96m";
const SIG: &str = "\x1b[90m";

fn base64url_decode(data: &str) -> Result<String, String> {
    let decoded = URL_SAFE_NO_PAD.decode(data).map_err(|e| e.to_string())?;
    String::from_utf8(decoded).map_err(|e| e.to_string())
}

fn colorize_value(v: &Value, indent: usize, level: usize) -> String {
    let spaces = " ".repeat(level * indent);
    match v {
        Value::Object(map) => {
            if map.is_empty() {
                return "{}".to_string();
            }
            let mut lines = vec!["{".to_string()];
            let len = map.len();
            let mut i = 0;
            for (k, val) in map {
                let key_str = format!("{}{}\"{}\"{}: ", KEY, RESET, k, RESET);
                let val_str = colorize_value(val, indent, level + 1);
                let mut line = format!("{}{}{}", spaces, " ".repeat(indent), key_str);
                line.push_str(&val_str);
                if i < len - 1 {
                    line.push(',');
                }
                lines.push(line);
                i += 1;
            }
            lines.push(format!("{}}}", spaces));
            lines.join("\n")
        }
        Value::Array(arr) => {
            if arr.is_empty() {
                return "[]".to_string();
            }
            let mut lines = vec!["[".to_string()];
            for (i, val) in arr.iter().enumerate() {
                let mut line = format!("{}{}{}", spaces, " ".repeat(indent), colorize_value(val, indent, level + 1));
                if i < arr.len() - 1 {
                    line.push(',');
                }
                lines.push(line);
            }
            lines.push(format!("{}]", spaces));
            lines.join("\n")
        }
        Value::String(s) => format!("{}{}\"{}\"{}", STR, RESET, s, RESET),
        Value::Bool(b) => format!("{}{}{}", BOOL, b, RESET),
        Value::Null => format!("{}null{}", NULL, RESET),
        Value::Number(n) => format!("{}{}{}", NUM, n, RESET),
    }
}

fn colorize_json(v: &Value, indent: usize) -> String {
    colorize_value(v, indent, 0)
}

fn decode_jwt(token: &str) -> Result<(Value, Value, String), String> {
    let parts: Vec<&str> = token.split('.').collect();
    if parts.len() != 3 {
        return Err("Неверный JWT: ожидается 3 сегмента".to_string());
    }
    let header_b64 = parts[0];
    let payload_b64 = parts[1];
    let sig = parts[2].to_string();
    let header_str = base64url_decode(header_b64)?;
    let payload_str = base64url_decode(payload_b64)?;
    let header: Value = serde_json::from_str(&header_str).map_err(|e| format!("Ошибка парсинга header: {}", e))?;
    let payload: Value = serde_json::from_str(&payload_str).map_err(|e| format!("Ошибка парсинга payload: {}", e))?;
    Ok((header, payload, sig))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut token = String::new();
    let mut file = None;
    let mut compact = false;

    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--file" => {
                if i + 1 < args.len() {
                    file = Some(args[i + 1].clone());
                    i += 2;
                } else { i += 1; }
            }
            "--compact" => {
                compact = true;
                i += 1;
            }
            _ => {
                if token.is_empty() {
                    token = args[i].clone();
                }
                i += 1;
            }
        }
    }
    if let Some(f) = file {
        token = fs::read_to_string(&f).expect("Ошибка чтения файла").trim().to_string();
    }
    if token.is_empty() {
        eprintln!("Не указан JWT-токен.");
        std::process::exit(1);
    }
    let (header, payload, sig) = match decode_jwt(&token) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("Ошибка: {}", e);
            std::process::exit(1);
        }
    };
    let indent = if compact { 0 } else { 2 };
    let mut stdout = StandardStream::stdout(ColorChoice::Auto);
    // Header
    stdout.set_color(ColorSpec::new().set_fg(Some(Color::Cyan))).unwrap();
    writeln!(&mut stdout, "--- Header ---").unwrap();
    stdout.reset().unwrap();
    println!("{}", colorize_json(&header, indent));
    // Payload
    stdout.set_color(ColorSpec::new().set_fg(Some(Color::Cyan))).unwrap();
    writeln!(&mut stdout, "--- Payload ---").unwrap();
    stdout.reset().unwrap();
    println!("{}", colorize_json(&payload, indent));
    // Signature
    stdout.set_color(ColorSpec::new().set_fg(Some(Color::White)).set_intense(false)).unwrap();
    writeln!(&mut stdout, "--- Signature ---").unwrap();
    stdout.reset().unwrap();
    println!("{}", sig);
}
